components/kyma-environment-broker/common/orchestration/strategies/parallel.go

```diff
@@ -0,0 +1,179 @@
+package strategies
+
+import (
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/kyma-project/control-plane/components/kyma-environment-broker/internal/storage/dberr"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"github.com/kyma-project/control-plane/components/kyma-environment-broker/common/orchestration"
+	"github.com/kyma-project/control-plane/components/kyma-environment-broker/internal/storage"
+
+	"github.com/pkg/errors"
+
+	"github.com/google/uuid"
+	"github.com/sirupsen/logrus"
+	"k8s.io/client-go/util/workqueue"
+)
+
+type Executor interface {
+	Execute(operationID string) (time.Duration, error)
+}
+
+type ParallelOrchestrationStrategy struct {
+	executor        Executor
+	dq              workqueue.DelayingInterface
+	orchestrations  storage.Orchestrations
```

Comment:
Strategy must be independent from storage. I'm about to raise the taskrun PR in CLI which leverages strategy and resolver. Please move all storage operations (cancellation checks) to the operation's Execute() or one of it's steps.

Reply:
I think we need to check somehow if orchestration wasn't canceled in case of `maintenanceWindow` schedule type. After adding it to the queue I'm not able to remove it so I decided to go with the polling storage function. Instead of using storage in strategies, can we just send a pointer to this waiter function and use different waiting strategies if it's provided or not? It would look like that: 
```
	if p.isCanceled != nil {
		waiter := *p.isCanceled
		isCanceled, err := waiter(until, log)
		if err != nil {
			return errors.Wrap(err, "while checking if orchestration was canceled")
		}
		if isCanceled {
			log.Info("Operation was canceled")
			return nil
		}
		log.Infof("Upgrade operation is scheduled now")
		p.dq.Add(id)
	} else {
		p.dq.AddAfter(id, until)
	}
```

Reply:
So in case of maintenancewindow the workers are blocked at `dq.Get()` . These internally used delaying queues cound only be cancelled with `Shutdown()` call which will signal and wake all blocked worker goroutines. So all you need is to store dq in strategy, and expose dq.Shutdown() call in strategy.Cancel() function.

Reply:
Didn't know that delaying queues have that fancy shutdown logic, exactly what I needed there thx!